pyteal/ast/abi/tuple.py

```diff
@@ -117,97 +122,128 @@ def _encode_tuple(values: Sequence[BaseType]) -> Expr:
     return Concat(*toConcat)
 
 
-def _index_tuple(
-    value_types: Sequence[TypeSpec], encoded: Expr, index: int, output: BaseType
-) -> Expr:
-    if not (0 <= index < len(value_types)):
-        raise ValueError("Index outside of range")
+@dataclass
+class _IndexTuple:
+    value_types: Sequence[TypeSpec]
+    encoded: Expr
 
-    offset = 0
-    ignoreNext = 0
-    lastBoolStart = 0
-    lastBoolLength = 0
-    for i, typeBefore in enumerate(value_types[:index]):
-        if ignoreNext > 0:
-            ignoreNext -= 1
-            continue
+    def __post_init__(self):
+        require_type(self.encoded, TealType.bytes)
 
-        if typeBefore == BoolTypeSpec():
-            lastBoolStart = offset
-            lastBoolLength = _consecutive_bool_type_spec_num(value_types, i)
-            offset += _bool_sequence_length(lastBoolLength)
-            ignoreNext = lastBoolLength - 1
-            continue
+    def get_or_store(self, index: int, output: BaseType | None = None) -> Expr:
+        if index not in range(len(self.value_types)):
```

Comment:
this seems like a slow way to check an index boundary

Reply:
pushed a faster way to check index boundary: [fast index boundary check](https://github.com/algorand/pyteal/pull/624/commits/9b56038e24ae76f51d427c3e0297c53273e5eb09)